core/templates/pages/story-editor-page/navbar/story-editor-navbar.component.spec.ts

```diff
@@ -106,20 +125,393 @@ describe('Story editor navbar directive', () => {
+      .and.returnValue(mockStoryInitializedEventEmitter);
+    spyOn(editableStoryBackendApiService, 'changeStoryPublicationStatusAsync')
+      .and.returnValue(Promise.resolve());
+
     component.ngOnInit();
-    expect(component.warningsAreShown).toEqual(false);
-    expect(component.forceValidateExplorations).toEqual(true);
-    expect(component.showNavigationOptions).toEqual(false);
-    expect(component.showStoryEditOptions).toEqual(false);
+    mockStoryInitializedEventEmitter.emit();
+    fixture.detectChanges();
+
+    expect(component.storyIsPublished).toBe(undefined);
+
+    storyEditorStateService.setStory(story);
+    fixture.detectChanges();
+    // This check will make sure that story is
+    // loaded correctly before publishing it.
+    expect(storyEditorStateService.hasLoadedStory()).toBe(true);
+
+    // Publishing story will acts as pre-check here.
+    component.publishStory();
+    tick(1000);
+    fixture.detectChanges();
+    expect(component.storyIsPublished).toBe(true);
+
+    component.unpublishStory();
+    tick(1000);
+    fixture.detectChanges();
+    expect(component.storyIsPublished).toBe(false);
+  }));
+
+  it('should toggle warning text', () => {
+    component.warningsAreShown = true;
+
+    component.toggleWarningText();
+    fixture.detectChanges();
+
+    expect(component.warningsAreShown).toBe(false);
+  });
+
+  it('should toggle navigation options', () => {
+    component.showNavigationOptions = true;
+
+    component.toggleNavigationOptions();
+    fixture.detectChanges();
+
+    expect(component.showNavigationOptions).toBe(false);
+  });
+
+  it('should toggle edit options', () => {
+    component.showStoryEditOptions = true;
+
+    component.toggleStoryEditOptions();
+    fixture.detectChanges();
+
+    expect(component.showStoryEditOptions).toBe(false);
+  });
+
+  it('should navigate to main tab in story editor page', () => {
+    expect(component.activeTab).toBe(undefined);
+
+    component.selectMainTab();
+
+    expect(component.activeTab).toBe('Editor');
+  });
+
+  it('should navigate to preview tab in story editor page', () => {
+    expect(component.activeTab).toBe(undefined);
+
+    component.selectPreviewTab();
+
+    expect(component.activeTab).toBe('Preview');
+  });
+
+  it('should discard changes', () => {
+    story = storyObjectFactory.createFromBackendDict(storyBackendDict);
+
+    component.story = story;
+    let clearChangesSpy = spyOn(
+      undoRedoService, 'clearChanges').and.returnValue();
+
+    component.discardChanges();
+
+    expect(clearChangesSpy).toHaveBeenCalled();
+  });
+
+  describe('open a confirmation modal for saving chages ', () => {
+    it('should save story successfully on' +
+      'clicking save draft button', fakeAsync(() => {
+      story = storyObjectFactory.createFromBackendDict(storyBackendDict);
+      let mockStoryInitializedEventEmitter = new EventEmitter();
+
+      spyOn(storyEditorStateService, 'getStory').and.returnValue(story);
```

Comment:
Can we move common spies to a beforeEach?

Reply:
Moved `editableStoryBackendApiService` to beforeEach,
I think we can not generalize all the spies since we are testing different scenarios in a story
so we need different stories in each case.